### ISSUE TITLE
AIF must be started after network is up causing a host to be exposed temporarily 

### ISSUE THREAD

**Issue by arnova** [arno-iptables-firewall/aif#41]
I just committed some stuff which changes AIF's behavior at start. Previously we would start before network was up which caused all kinds of problems with eg. plugins which need to setup routes or try to resolve names etc. I've now changed the service file so that AIF is started right after network is up, but this leaves the host exposed temporarily. This window of opportunity is very small but we *may* want to fix this by eg. adding a helper service which simply inits all builtin iptables chains to DROP.

**Comment by arnova** [comment]
@abelbeck : Ideas? How is this doen on Ast? It probably doesn't use systemd, does it?

**Comment by abelbeck** [comment]
@arnova In AstLinux we have always brought up the network stack slightly before AIF, but no network services are started before AIF.

I would think as long as AIF is started before any network services are started, we are good.

Possibly some kernel default sysctl's settings could generate network traffic before AIF is started, mostly IPv6 RA stuff (ICMPv6), we build our kernel with `ipv6` as a module and only load it as needed:
```
    modprobe ipv6
    # Disable IPv6 Autoconf by default
    IFS=$'\n'
    for line in $(sysctl -a 2>/dev/null | cut -d' ' -f1 | \
        grep -e "^net\.ipv6\.conf\..*\.autoconf$" \
             -e "^net\.ipv6\.conf\..*\.accept_ra$"); do
      sysctl -w "$line=0" >/dev/null
    done
```
These `sysctl` entries are addressed again in AIF.  Though Debian has `ipv6` builtin the kernel, not as a module.

AstLinux does not use systemd (insert happy-dance emoji) :-)

**Comment by arnova** [comment]
Well, that's the problem: Several (network) services are already started, even before the network is up. So as as soon as the network is up those services are accessible.

**Comment by abelbeck** [comment]
That's not ideal.

The good news is AIF should block any further (disallowed) traffic that may have started in the network-up -> AIF startup window.  Leaving very little time to brute force a SSH password.

I suppose for systemd systems you could add a helper service as you mentioned ... does the iptables package default to any rules ?

Does Debian expect DNS to work before AIF is started ?  If so a universal DROP in a helper service may cause problems.

**Comment by arnova** [comment]
Honestly I have no idea what Debian expects. I did notice that on one of my machines starting AIF too early causes major issues, also since /var is mounted on a separate partition which wasn't properly handled before either. Ideally one would like to have this chain of events: network up, AIF started, network daemons started. I'll do some more thinking about the issue. For now at least AIF no longer breaks "certain" systems which seems the most important fix.

**Comment by abelbeck** [comment]
> Ideally one would like to have this chain of events: network up, AIF started, network daemons started. For now at least AIF no longer breaks "certain" systems which seems the most important fix.

I agree.

**Comment by flixman** [comment]
In Systemd, isn't that what [network-pre.target](https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/) is for? 

    network-pre.target is a target that may be used to order services before any network interface is  
    configured. Its primary purpose is for usage with firewall services that want to establish a firewall 
    before any network interface is up. It's a passive unit: you cannot start it directly and it is not 
    pulled in by the the network management service, but by the service that wants to run before it. 
    Network management services hence should set After=network-pre.target, but avoid any 
    Wants=network-pre.target or even Requires=network-pre.target. Services that want to be run 
    before the network is configured should place Before=network-pre.target and also set 
    Wants=network-pre.target to pull it in. This way, unless there's actually a service that needs to be 
    ordered before the network is up the target is not pulled in, hence avoiding any unnecessary 
    synchronization point.

I understand that the dependencies should be, then

    Before=network-pre.target
    Wants=network-pre.target
    After=nftables.service

**Comment by arnova** [comment]
Keep in mind that when the network is up but the firewall isn't yet, there are no services running yet so the security impact is minimal. Another reason is that there were issues on at least older Debian versions where systemd would sometimes "choke" when the firewall was started before the network interface(s) were brought up.

**Comment by flixman** [comment]
@arnova  "Keep in mind that when the network is up but the firewall isn't yet, there are no services running yet so the security impact is minimal." Are you sure on this? what I see is that, e.g., both aif, dovecot, rspamd, etc., do have directive "After= [...] network-online.target", which means that all those services might be started concurrently. Therefore, there might be services started before the firewall comes into place.

In my debian, I see that there many services (e.g. apache web server) started between network.target and network-online.target, which suggests that aif should be started before network.target and after the network interfaces are up... but I have not been able to find such unit.

Maybe this could be added as a script, that checks the configuration files and sets the dependencies of aif with respect to the systemd's network .device units?

**Comment by arnova** [comment]
Yes, now I remember. Like you already figured out yourself, there's no optimal way to implement this. Hacking configuration files etc. is NOT the way to go so for now we'll have to accept this sub-optimal solution. Keep in mind that the "window-of-opportunity" is still very, very small.
